virtual_rainforest/core/core_components.py

```diff
@@ -0,0 +1,278 @@
+    subcanopy_layer_height: float = field(init=False)
+    """The height above ground used to represent subcanopy conditions."""
+    layer_roles: tuple[str, ...] = field(init=False)
+    """An ordered tuple giving the roles of the layers within the model."""
+    config: InitVar[Config]
+    """A validated model configuration."""
+
+    def __post_init__(self, config: Config) -> None:
+        """Populate the ``LayerStructure`` instance.
+
+        This method populates the ``LayerStructure`` attributes from the provided
+        :class:`~virtual_rainforest.core.config.Config` instance.
+
+        Args:
+            config: A Config instance.
+        """
+
+        lyr_config = config["core"]["layers"]
+
+        # Validate contents
+        # NOTE: some of this is also trapped by validation against the core schema.
+        # Canopy layers
+        canopy_layers = lyr_config["canopy_layers"]
+        if (
+            not isinstance(canopy_layers, int)
+            and not (isinstance(canopy_layers, float) and canopy_layers.is_integer())
+            and canopy_layers < 1
+        ):
+            to_raise = ConfigurationError(
+                "The number of canopy layers is not a positive integer."
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        self.canopy_layers = canopy_layers
+
+        # Soil layers
+        soil_layers = lyr_config["soil_layers"]
+
+        if not isinstance(soil_layers, list) or len(soil_layers) < 1:
+            to_raise = ConfigurationError(
+                "The soil layers must be a non-empty list of layer depths."
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        if not all([isinstance(v, (float, int)) for v in soil_layers]):
+            to_raise = ConfigurationError("The soil layer depths are not all numeric.")
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        np_soil_layer = np.array(soil_layers)
+        if not (np.all(np_soil_layer < 0) and np.all(np.diff(np_soil_layer) < 0)):
+            to_raise = ConfigurationError(
+                "Soil layer depths must be strictly decreasing and negative."
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        self.soil_layers = soil_layers
+
+        # Other heights should all be positive floats
+        for attr, value in (
+            ("above_canopy_height_offset", lyr_config["above_canopy_height_offset"]),
+            ("surface_layer_height", lyr_config["surface_layer_height"]),
+            ("subcanopy_layer_height", lyr_config["subcanopy_layer_height"]),
+        ):
+            if not isinstance(value, (float, int)) or value < 0:
+                to_raise = ConfigurationError(
+                    f"The {attr} value must be a positive float."
+                )
+                LOGGER.error(to_raise)
+                raise to_raise
+            else:
+                setattr(self, attr, value)
```

Comment:
And the same here:

```python
def _validate_positive(value) -> int | float:
    ...

setattr(self, attr, _validate_positive(value))

Reply:
Things like this might be a useful wider `util` - we should probably do a code base scan at some point looking for duplicated patterns like this.

Reply:
@dalonsoa Turns out `np.infty` and `np.nan` both sail straight through `isinstance(value, (float, int))`, so we now have `_validate_positive_finite_numeric` 😄